### PR TITLE
[crossgen2] Avoid ambiguous devirtualization

### PR DESCRIPTION
**Pull request by cshung**
It is possible for us to hit a scenario where a type implements the same interface more than once due to generic instantiations.

In some instances of those cases, the [VirtualMethodAlgorithm ](https://github.com/dotnet/coreclr/blob/master/src/tools/crossgen2/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs)does not produce identical output as CoreCLR would, leading to behavioral differences in compiled outputs.

A canonical example for [this](https://github.com/dotnet/coreclr/blob/f1a900df6554a682b2816b64958f80055407e955/tests/src/Loader/classloader/InterfaceFolding/Ambiguous.il#L141) is interface dispatch.

Instead of fixing the algorithm (in which the work to fix it is tracked in https://github.com/dotnet/corert/issues/208), the PR introduces an early duplication detection algorithm will detect the case and refuse to devirtualize for those scenarios.

@dotnet/crossgen-contrib

### CODE REVIEW

File: src/tools/crossgen2/Common/Compiler/DevirtualizationManager.cs [hunk 1]
```diff
@@ -74,6 +74,32 @@ protected virtual MethodDesc ResolveVirtualMethod(MethodDesc declMethod, DefType
             if (declMethod.OwningType.IsInterface)
             {
                 impl = implType.ResolveInterfaceMethodTarget(declMethod);
+
```
Comment (MichalStrehovsky) [inline src/tools/crossgen2/Common/Compiler/DevirtualizationManager.cs]: Could you put this into src\tools\crossgen2\ILCompiler.ReadyToRun\Compiler\DependencyAnalysis\ReadyToRun\DevirtualizationManager.cs instead (right before it calls into `base`)?

Being able to work around like this is ready to run specific.

**Comment by sergiy-k** [comment]
@cshung, may I please ask you to make sure that we have a .NET 5 issue in the CoreCLR repo to track the proper solution?

**Comment by MichalStrehovsky** [comment]
> @cshung, may I please ask you to make sure that we have a .NET 5 issue in the CoreCLR repo to track the proper solution?

@davidwrighton can correct me if I'm wrong here, but we shipped .NET Native with this bug (I think Reducer was messing this up long before it hit NUTC). We had the bug open for a long time, it never met the bar, and then we just deleted the tests and closed the bugs. I believe this can only be hit with IL. Chances are the new issue will be met with the same fate (it even matters less for R2R, because it only affects an optimization for the corner case, not correctness).

**Comment by sergiy-k** [comment]
If this can only be hit as a corner case with IL, then I agree - no need further invest into this issue. Thank you!

**Comment by davidwrighton** [comment]
Agreed. Since this change prevents the issue from being a correctness issue, I do not expect we'll ever have a good reason to do the very risky work of attempting to "fix" the issue in the core virtual resolution algorithm.

**Pull request merged**